crates/bevy_ecs/src/schedule/condition.rs

```diff
@@ -915,11 +922,117 @@ pub mod common_conditions {
     /// app.run(&mut world);
     /// assert_eq!(world.resource::<Counter>().0, 0);
     /// ```
-    pub fn not<Marker, T>(condition: T) -> impl Condition<()>
+    pub fn not<Marker, T>(condition: T) -> ConditionInverter<Marker, T::System>
     where
+        Marker: 'static,
         T: Condition<Marker>,
     {
-        condition.pipe(|In(val): In<bool>| !val)
+        let condition = IntoSystem::into_system(condition);
+        let name = format!("!{}", condition.name());
+        ConditionInverter::<Marker, T::System> {
+            _marker: PhantomData,
+            condition,
+            name: Cow::Owned(name),
+        }
+    }
+}
+
+/// Inverts the output of a [`Condition`].
+///
+/// See [`common_conditions::not`] for examples.
+
+pub struct ConditionInverter<Marker, T>
+where
+    T: System<In = (), Out = bool> + ReadOnlySystem,
```

Comment:
```suggestion
    T: System<In = (), Out = bool>,
```

I don't think anything bad would happen if non-read-only systems were used with this type.

Reply:
I don't think we can safely implement `ReadOnlySystem` if we can't be sure the inner system isn't also read-only?

Reply:
That bound only needs to be on `impl ReadOnlySystem for ConditionInverter`. It's unnecessary on the type itself, and the rest of the impls.